### PR TITLE
Add linker for Hexagon, use shared objects on all targets

### PR DESCRIPTION
**Pull request by dsharletg**
This is a large PR, which is a breaking change for the Hexagon runtime.

Currently, Hexagon offloaded pipelines are compiled in two ways: on 8996 devices, they are compiled to an object, and loaded with @abadams's heroic object loader. On 8998 devices, mmap/mprotect are going away, and so we can't use that loader anymore. Instead, on 8998 devices, we need to build a shared object, and load it with dlopen on 8998. 

This is problematic for a few reasons:

- It means that to support 8996 and 8998, two versions of the same pipeline are needed: one compiled to an object, one compiled to a shared object (with the hvx_shared_object target flag).
- It shells out to hexagon-clang to link the shared object. This is problematic for some build systems.
- On 8996, we are in danger of coming across new code that generates a relocation type we don't support. Because this code is in the runtime, rather than the compiler, this is a much more risky situation (fixing some missing support is much easier to do in the compiler than it is to do in the runtime which is distributed to many devices).

At first, I hoped to implement a linker that would generate a binary that is both a valid object and a valid shared object. However, our object loader is missing crucial relocation support for this to work. So, we've decided that we need a breaking change in the Hexagon runtime, which is that we are always supporting shared objects only. To continue supporting 8996, this PR modifies the object loader to load shared objects instead. The runtime uses dlopenbuf when it is available, and uses the (now shared) object loader when not available.

I tried to build the linker in a way that isolates the platform specific code. I'm hoping that we could extend it to support x86 as well, so we could use it to generate matlab mex binaries directly, instead of relying on matlab's mex command. x86 appears to have very simple relocations, so this might not be hard to do.

To summarize, this PR makes the following changes:

- HexagonOffload always builds and embeds shared objects in the pipeline.
- The runtime always loads shared objects.
- The runtime uses dlopenbuf when it is available, and falls back to a custom dlopen implementation based on mmap/mprotect when it is unavailable. This should enable us to deprecate the custom mmap/mprotect dlopen implementation in the future.

### CODE REVIEW

**Comment by ronlieb** [comment]
Minor cleanup needed for clang++ compilation
Elf.cpp has a few signed unsigned compares  lines 372 and 400

**Comment by ronlieb** [comment]
compiles and runs nicely (yay), still needs fopen -> open patch applied

File: src/HexagonOffload.cpp [hunk 1]
```diff
@@ -12,13 +12,554 @@
...
 
 namespace Halide {
 namespace Internal {
 
 using std::string;
 using std::vector;
 
+namespace Elf {
+
+enum {
```
Comment (steven-johnson) [inline src/HexagonOffload.cpp]: Is it worth adding a comment re: the source of these constants, for future reference?

**Comment by dsharletg** [comment]
I just merged master, PTAL

File: src/Module.cpp [hunk 2]
```diff
@@ -227,30 +227,7 @@ Module link_modules(const std::string &name, const std::vector<Module> &modules)
...
-            llvm::raw_svector_ostream assembly_stream(assembly);
-            compile_llvm_module_to_assembly(*llvm_module, assembly_stream);
-            debug(2) << assembly.c_str() << "\n";
-        }
-
-        Buffer<uint8_t> result(object.size(), name());
-        memcpy(result.data(), reinterpret_cast<uint8_t*>(&object[0]), object.size());
-        return result;
-    }
+    return compile_module_to_hexagon_shared_object(*this);
```
Comment (dsharletg) [inline src/Module.cpp]: @zvookin you might want to look at this in particular. I think we need "compile_to_shared_object" to be more general (not Hexagon specific), and a mechanism for submodules to specify how they want to be compiled. For another PR...

File: src/Elf.h [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+#include <list>
+#include <string>
+#include <iterator>
+#include <limits>
+
+namespace Halide {
+namespace Internal {
+namespace Elf {
+
+// This ELF parser mostly deserializes the object into a graph
```
Comment (abadams) [inline src/Elf.h]: If you have an ELF spec or something that documents the meaning of things like STB_GLOBAL, please add a link to it here.

File: src/HexagonOffload.cpp [hunk 4]
```diff
@@ -283,12 +825,13 @@ class InjectHexagonRpc : public IRMutator {
...
 };
 
-}
+}  // namespace
 
 Stmt inject_hexagon_rpc(Stmt s, const Target &host_target,
                         Module &containing_module) {
     // Make a new target for the device module.
     Target target(Target::NoOS, Target::Hexagon, 32);
+    target = target.with_feature(Target::NoAsserts);
```
Comment (abadams) [inline src/HexagonOffload.cpp]: Is this an intentional change?

File: src/runtime/hexagon_remote/dlib.cpp [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+                return sym;
+            }
+            i = hash.next_in_chain(i);
+        }
+        return NULL;
+    }
+};
+
+// dlopen a relocatable (but not yet relocated) object file in
+// memory. The object should be compiled with -fno-pic.
```
Comment (abadams) [inline src/runtime/hexagon_remote/dlib.cpp]: Is this comment still accurate?

File: src/Elf.h [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+    ///@}
+
+    /** Access the binding and visibility of this symbol. See the ELF
+     * spec for more information about these properties. */
+    ///@{
+    Symbol &set_binding(Binding binding) {
+        this->binding = binding;
+        return *this;
+    }
+    Symbol &set_visibility(Visibility visibility) {
```
Comment (abadams) [inline src/Elf.h]: This file has a mix of single-line implementations and multi-line ones, even for the setters. I'd just make them all multi-line.

File: src/Elf.h [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+    Section &set_contents(const std::vector<T> &contents) {
+        this->contents.assign((const char *)contents.data(), (const char *)(contents.data() + contents.size()));
+        return *this;
+    }
+    template <typename T>
+    Section &append_contents(const T& x) {
+        return append_contents((const char *)&x, (const char *)(&x + 1));
+    }
+    const std::vector<char> &get_contents() const { return contents; }
+    contents_iterator contents_begin() { return contents.begin(); }
```
Comment (abadams) [inline src/Elf.h]: Is all this iterating functionality used? Might be simpler to just have contents()/relocations() return a reference to the underlying vector.

File: src/runtime/hexagon_remote/dlib.cpp (hunk 5 above)
Comment (dsharletg) [inline src/runtime/hexagon_remote/dlib.cpp]: Updated and moved the comments to the header.

File: src/HexagonOffload.cpp (hunk 4 above)
Comment (dsharletg) [inline src/HexagonOffload.cpp]: No, good catch!

File: src/Elf.h (hunk 7 above)
Comment (dsharletg) [inline src/Elf.h]: I think a lot of it is. The motivation for this is that when I was working on it, I kept going back and forth between vector and list, eventually settling on list (so iterators aren't invalidated when adding/removing items to the container). We still might want to change this in the future.

Comment (dsharletg) [inline src/Elf.h]: It's also how LLVM's object file API looks, and I was trying to use that for a long time before giving up (it's fine for parsing, not good for modifying).

File: src/Elf.h (hunk 6 above)
Comment (dsharletg) [inline src/Elf.h]: I made all of the multi-statement setters multi-line, left the single statement things (getters) single line.

**Comment by abadams** [comment]
Minor windows breakage: 

https://buildbot.halide-lang.org/master/builders/win-64-testbranch-400/builds/2/steps/Build%20Halide%20Release/logs/stdio

**Comment by abadams** [comment]
Oh, not just windows: https://buildbot.halide-lang.org/master/builders/mac-32-testbranch-400/builds/1/steps/make%20distrib/logs/stdio

File: src/Elf.cpp [hunk 8]
```diff
@@ -0,0 +1,989 @@
...
+
+    static addr_t make_info(uint32_t type, uint32_t sym) {
+        if (sizeof(addr_t) == 8) {
+            return (uint64_t)type | ((uint64_t)sym << 32);
+        } else {
+            return (type & 0xff) | (sym << 8);
+        }
+    }
+
+    void set_r_type(uint32_t type) {
```
Comment (shankarke) [inline src/Elf.cpp]: set_type.

File: src/Elf.cpp [hunk 9]
```diff
@@ -0,0 +1,989 @@
...
+        r_info = make_info(type, r_sym());
+    }
+
+    void set_r_sym(uint32_t sym) {
```
Comment (shankarke) [inline src/Elf.cpp]: set_sym.

File: src/Elf.cpp [hunk 10]
```diff
@@ -0,0 +1,989 @@
...
+
+    Rela(addr_t offset, uint32_t type, uint32_t sym, addr_off_t addend)
+        : Rel<T>(offset, type, sym), r_addend(addend) {}
+};
+
+template <typename T>
+struct Sym;
+
+template <>
+struct Sym<Types<32>> {
```
Comment (shankarke) [inline src/Elf.cpp]: Are you using the template here ?

File: src/Elf.cpp [hunk 11]
```diff
@@ -0,0 +1,989 @@
...
+        uint32_t d_val;
+        addr_t d_ptr;
+    };
+};
+
+class StringTable {
+    // TODO: We could be smarter and find substrings in the existing
+    // table, not just whole strings. It would probably be fine to just
+    // put every substring of each new string into the cache.
+    std::map<std::string, uint32_t> cache;
```
Comment (shankarke) [inline src/Elf.cpp]: std::unordered_map ?

File: src/Elf.cpp [hunk 12]
```diff
@@ -0,0 +1,989 @@
...
+        if (i != symbols.end()) return i->second;
+        return sym;
+    };
+
+    // Check if a relocation needs a PLT entry, which adds some
+    // additional conditions on top of what the linker implementation
+    // wants.
+    auto needs_plt_entry = [&](const Relocation &r) {
+        const Symbol *s = get_symbol(r);
+        if (!s || s->is_defined()) {
```
Comment (shankarke) [inline src/Elf.cpp]: no local symbols need a PLT entry.

symbols that can be internally bound, need not have a PLT entry.

File: src/Elf.cpp [hunk 13]
```diff
@@ -0,0 +1,989 @@
...
+            plt_def = &plt_symbols.back();
+            symbols[plt_def] = plt_def;
+        }
+    }
+
+    // Start placing the sections into the shared object.
+
+    // Leave room for the header, and program headers at the beginning of the file.
+    append_zeros(output, sizeof(ehdr));
+    append_zeros(output, sizeof(phdrs[0])*3);
```
Comment (shankarke) [inline src/Elf.cpp]: I am not sure whats 3 here ? Is it one PT_LOAD for text, PT_LOAD for data and PT_DYNAMIC ?

Is there an assumption, if so, we should list that.

File: src/Elf.cpp [hunk 14]
```diff
@@ -0,0 +1,989 @@
...
+    symtab.set_flag(Section::SHF_ALLOC);
+    std::vector<Sym<T>> syms;
+    Sym<T> undef_sym;
+    memset(&undef_sym, 0, sizeof(undef_sym));
+    syms.push_back(undef_sym);
+    // TODO: This map with pointers as the key will lead to non-deterministic builds...
+    std::map<const Symbol *, uint16_t> symbol_idxs;
+    uint64_t local_count = 0;
+    for (bool is_local : {true, false}) {
+        for (const auto &i : symbols) {
```
Comment (shankarke) [inline src/Elf.cpp]: You might not want to add any local symbol inside the symbol table et all. Since its the dynamic symbol table. This way there would be less confusion and makes the code much simpler too.

File: src/Elf.cpp [hunk 15]
```diff
@@ -0,0 +1,989 @@
...
+        if (is_local) {
+            local_count = syms.size();
+        }
+    }
+    symtab.set_contents(syms);
+    uint16_t symtab_idx = write_section(symtab, sizeof(syms[0]));
+    safe_assign(shdrs[symtab_idx].sh_info, local_count);
+
+    // Also write the symbol table as SHT_DYNSYM.
+    Section dynsym = symtab;
```
Comment (shankarke) [inline src/Elf.cpp]: For shared objects, the symbol table is useless, you can save on memory space by just creating dynsym.

File: src/Elf.cpp [hunk 16]
```diff
@@ -0,0 +1,989 @@
...
+    strings.get(soname);
+    for (const auto &i : dependencies) {
+        strings.get(i);
+    }
+
+    Section dynamic(".dynamic", Section::SHT_DYNAMIC);
+    strings.get(dynamic.get_name());
+    dynamic.set_alignment(4);
+    dynamic.set_flag(Section::SHF_ALLOC);
+    Section strtab(".strtab", Section::SHT_STRTAB);
```
Comment (shankarke) [inline src/Elf.cpp]: The strtab for dynamic sections is called .dynstr.

File: src/Elf.cpp [hunk 17]
```diff
@@ -0,0 +1,989 @@
...
+        d.d_tag = tag;
+        d.d_val = val;
+        return d;
+    };
+
+    for (const auto &i : dependencies) {
+        dyn.push_back(make_dyn(DT_NEEDED, strings.get(i)));
+    }
+    if (!soname.empty()) {
+        dyn.push_back(make_dyn(DT_SONAME, strings.get(soname)));
```
Comment (shankarke) [inline src/Elf.cpp]: For Halide, soname is mostly empty right.

File: src/HexagonOffload.cpp [hunk 18]
```diff
@@ -12,13 +12,558 @@
...
+    }
+}
+
+std::string hex(uint32_t x) {
+    char buffer[100];
+    snprintf(buffer, sizeof(buffer), "0x%08x", x);
+    return buffer;
+}
+
+void do_reloc(char *addr, uint32_t mask, uintptr_t val, bool is_signed, bool verify) {
```
Comment (shankarke) [inline src/HexagonOffload.cpp]: You might want to use the most recent relocation changes that made it to Hexagon. This code can be completely removed when you do that.

File: src/Elf.cpp [hunk 19]
```diff
@@ -0,0 +1,989 @@
...
+
+    // Other relocations.
+    dyn.push_back(make_dyn(DT_RELA, shdrs[rela_got_idx].sh_offset + pltrelsz));
+    dyn.push_back(make_dyn(DT_RELASZ, shdrs[rela_got_idx].sh_size - pltrelsz));
+    dyn.push_back(make_dyn(DT_RELAENT, sizeof(Rela<T>)));
+
+    dynamic.set_contents(dyn);
+
+    // Add any target specific stuff.
+    linker->append_dynamic(dynamic);
```
Comment (shankarke) [inline src/Elf.cpp]: Placing sections in the right order is important, to assign virtual addresses.

File: src/Elf.cpp [hunk 20]
```diff
@@ -0,0 +1,989 @@
...
+    uint32_t sh_name;
+    uint32_t sh_type;
+    addr_t sh_flags;
+    addr_t sh_addr;
+    addr_t sh_offset;
+    addr_t sh_size;
+    uint32_t sh_link;
+    uint32_t sh_info;
+    addr_t sh_addralign;
+    addr_t sh_entsize;
```
Comment (dsharletg) [inline src/Elf.cpp]: Hmm, it looks like this is a type that is the same as uintptr_t (which is addr_t), looking at Elf32_shdr and Elf64_shdr here: https://github.com/torvalds/linux/blob/master/include/uapi/linux/elf.h#L312

Still, I will clean this up a bit.

We can't include the linux header because this needs to build on windows.

File: src/Elf.cpp (hunk 10 above)
Comment (dsharletg) [inline src/Elf.cpp]: It's a specialization, I haven't bothered to define the 64 bit version yet. Will add a comment.

File: src/Elf.cpp (hunk 14 above)
Comment (dsharletg) [inline src/Elf.cpp]: I was adding these so the disassembly was more readable, I noticed that without the local symbols, the disassembly was just a giant blob. It's fuzzy to be now, but I think the behavior I was seeing is that the disassembly used the .symtab to annotate the assembly with the symbols, and not the .dynsym. I think the thing I should do is make a .symtab with all the symbols, and the a .dynsym with just the linker relevant symbols.

File: src/Elf.cpp (hunk 17 above)
Comment (dsharletg) [inline src/Elf.cpp]: Right, soname is empty by default.

File: src/Elf.cpp (hunk 19 above)
Comment (dsharletg) [inline src/Elf.cpp]: Is this order wrong? I noticed that the hexagon linker put the dynamic section early, but that requires writing it before some of the later addresses are known. It didn't seem like it was breaking anything to just put it at the end?

File: src/HexagonOffload.cpp (hunk 18 above)
Comment (dsharletg) [inline src/HexagonOffload.cpp]: Do you mean that this relocation handling code is now available in LLVM? A quick googling hasn't turned up the relevant code yet, but we would definitely prefer to depend on LLVM to do this... Ideally, we would use LLVM to do the whole linker job, we would love to get rid of this custom linker if possible.
